Assets/MRTK/Core/Utilities/Rendering/MaterialInstance.cs

```diff
@@ -146,7 +155,7 @@ private void Awake()
         private void Update()
         {
             // If the materials get changed via outside of MaterialInstance.
-            var sharedMaterials = CachedRenderer.sharedMaterials;
+            var sharedMaterials = CachedSharedMaterials;
```

Comment:
Not being part of the MRTK Team, I don't really understand why would we want to check whether materials has been changed outside of the MaterialInstance API. If this is Editor-only, then maybe this code should be sorrounded by #if preprocessor directive?

Reply:
This is a great question. Ideally we wouldn't worry about developers changing materials outside of the MaterialInstance API. But, I worried that some developers wouldn't be aware of the existence of the MaterialInstance and blindly assign to the renderer's shared material (causing the whole system to fall apart). 

As you point out this workflow is unavoidable in the editor, so maybe it's best to just wrap this in an editor preprocessor and fail when a developer assigns to the renderer's shared materials at runtime? @MaxWang-MS do you have any thoughts? We could also wrap this check in a `DEBUG` preprocessor so it runs in editor and in development builds and runs unchecked in release?

Reply:
I think it is likely that someone not very familiar with MRTK will make the mistake of modifying shared material directly instead of going through the material instance class, so I think we should either keep we what we had before or add a default-to-false switch that enables this optimization. I am leaning towards the former unless calling render.sharedMaterials is actually expensive.

Reply:
I agree with the former since much of the MRTK still doesn't track instance materials appropriately. Maybe we can make this change after #6166 is addressed?

Reply:
Just to put my two cents: iirc, after replacing renderer.sharedMaterials calls with this cache we have gained around 0.5ms and removed some garbage allocation. As you know, in AR environment it's a lot. Renderer.sharedMaterials useses some internal extern method that caught my attention while analyzing the Profiler.

Reply:
I agree anytime we reach into a container held by the engine we incur a bunch of overhead moving and copying from native to managed. I hate to complicate this, but maybe we make it an "opt-in" optimization for users that need it? Else, the behavior is the same for existing users?

Reply:
Hi @Cameron-Micka, I made this optimization opt-in as you suggested. Please let me know if it looks good now.